jobserver/views/users.py

```diff
@@ -156,6 +165,81 @@ def login_invalid(self):
         return redirect("login")
 
 
+class LoginWithToken(View):
+    def post(self, request):
+
+        form = TokenLoginForm(request.POST)
+        if not form.is_valid():
+            return self.login_invalid(form)
+
+        user_value = form.cleaned_data["user"]
+
+        try:
+            # username or email
+            user = User.objects.get(
+                Q(email=user_value)
+                | Q(username=user_value)
+                | Q(notifications_email=user_value)
+            )
+
+            # only github users can log in with token
+            if not user.social_auth.exists():
+                raise InvalidTokenUser(f"{user_value} is not a github user")
+
+            user.validate_login_token(form.cleaned_data["token"])
+
+        except (
+            User.DoesNotExist,
+            InvalidTokenUser,
+            User.BadLoginToken,
+            User.ExpiredLoginToken,
+        ) as e:
+            logger.exception(f"Login with token failed for user {user_value}")
+            trace.get_current_span().record_exception(e)
+            return self.login_invalid(form)
+
+        login(self.request, user, "django.contrib.auth.backends.ModelBackend")
+        messages.success(
+            self.request,
+            "You have been logged in using a single use token. That token is now invalid.",
+        )
+
+        next_url = get_next_url(self.request)
+        return redirect(next_url)
+
+    def login_invalid(self, form):
+        messages.error(
+            self.request,
+            "Login failed. The user or token may be incorrect, or the token may have expired.",
+        )
+        return TemplateResponse(
+            self.request,
+            template="login.html",
+            context={
+                "token_form": form,
+                "next_url": get_next_url(self.request),
+            },
+        )
+
+
+@method_decorator(login_required, name="dispatch")
+class GenerateLoginToken(View):
+    template = "generate-login-token.html"
+
+    def get(self, request):
+        return TemplateResponse(self.request, self.template)
+
+    def post(self, request):
```

Comment:
Can you add `, *args, **kwargs` to these handler invocations?  Keeps them consistent with the rest of the CBVs 🙂 (path of least surprise etc)

Reply:
done